### DISCUSSION TITLE
Float libraries undefined, "powf", "logf", and many more

### DISCUSSION THREAD

**Discussion by synapticonashaikh** [micropython/micropython#11412] [Libraries & Drivers] [answered]
Hello, I'm working on a project in which the processor is by XMOS, and as of now MicroPython doesn't support the processor, so I've modified the minimal port by replacing the HAL layer to be able to work for it. To do so, I've used the sample code of other ports (STM and Microchip) as a reference and developed the HAL layer myself (timer, gpio and other modules).

I've also modified the compilation process and made it very straightforward (updated the makefile a little).
`xcc -target=targent_name some_compilation_flags all_user_source_file complete_py_library -o output.xe`
The user source file is nothing but `main.c, _frozen_mpy.c` and some HAL modular files. It is working perfectly fine. 

Now I'm building an advanced application to support the overall software, where I want to do a cross-compile to the same code for the Linux environment. So I can use it for the simulation of the script (also testing of the modular library). I tried to replace the xcc (XMOS compiler) with cc and hope it will work without any hustle (because the entire code is still simple c, I guess?), but I'm getting lots of undefined references regards to float libraries. So, do I need to add some support compilation flags to support the library? 

Could anyone help me with this issue? Some error codes are mentioned down below:

Errors: 
```
/usr/bin/ld: /tmp/cca42kkT.o: in function `mp_format_float':
formatfloat.c:(.text+0x1a9): undefined reference to `powf'
/usr/bin/ld: formatfloat.c:(.text+0x1e3): undefined reference to `powf'
/usr/bin/ld: formatfloat.c:(.text+0x262): undefined reference to `powf'
/usr/bin/ld: formatfloat.c:(.text+0x2a0): undefined reference to `powf'
/usr/bin/ld: formatfloat.c:(.text+0x397): undefined reference to `powf'
/usr/bin/ld: /tmp/cclwvsAY.o:modbuiltins.c:(.text+0x709): more undefined references to `powf' follow
/usr/bin/ld: /tmp/cclwvsAY.o: in function `mp_builtin_round':
modbuiltins.c:(.text+0x721): undefined reference to `nearbyintf'
/usr/bin/ld: modbuiltins.c:(.text+0x73b): undefined reference to `nearbyintf'
/usr/bin/ld: /tmp/ccb6aPyy.o: in function `complex_unary_op':
objcomplex.c:(.text+0x200): undefined reference to `sqrtf'
/usr/bin/ld: /tmp/ccb6aPyy.o: in function `mp_obj_complex_binary_op':
objcomplex.c:(.text+0x5a2): undefined reference to `sqrtf'
/usr/bin/ld: objcomplex.c:(.text+0x614): undefined reference to `logf'
/usr/bin/ld: objcomplex.c:(.text+0x631): undefined reference to `atan2f'
/usr/bin/ld: objcomplex.c:(.text+0x66b): undefined reference to `sincosf'
/usr/bin/ld: objcomplex.c:(.text+0x6a7): undefined reference to `expf'
```

**Comment by synapticonashaikh** [comment, marked as answer]
Hello, I found one solution here on StackOverflow:
https://stackoverflow.com/questions/8671366/undefined-reference-to-pow-and-floor
We must add the `-lm` flag after all the source code files.

**Reply by jimmo** [reply]
That will use the compiler's math library. You can also use the one included with MicroPython (which is what the STM32 port does).  Search for `LIBM_SRC_C` in `stm32/Makefile`.

**Reply by synapticonashaikh** [reply]
What is the difference between these two libraries? Is the one with MicroPython more efficient and consumes less space?
